mart/attack/adversary.py

```diff
@@ -22,329 +25,171 @@
     from .objective import Objective
     from .perturber import Perturber
 
-__all__ = ["Adversary", "Attacker"]
-
-
-class AttackerCallbackHookMixin(Callback):
-    """Define event hooks in the Adversary Loop for callbacks."""
-
-    callbacks = {}
-
-    def on_run_start(self, **kwargs) -> None:
-        """Prepare the attack loop state."""
-        for _name, callback in self.callbacks.items():
-            # FIXME: Skip incomplete callback instance.
-            # Give access of self to callbacks by `adversary=self`.
-            callback.on_run_start(**kwargs)
-
-    def on_examine_start(self, **kwargs) -> None:
-        for _name, callback in self.callbacks.items():
-            callback.on_examine_start(**kwargs)
-
-    def on_examine_end(self, **kwargs) -> None:
-        for _name, callback in self.callbacks.items():
-            callback.on_examine_end(**kwargs)
-
-    def on_advance_start(self, **kwargs) -> None:
-        for _name, callback in self.callbacks.items():
-            callback.on_advance_start(**kwargs)
-
-    def on_advance_end(self, **kwargs) -> None:
-        for _name, callback in self.callbacks.items():
-            callback.on_advance_end(**kwargs)
-
-    def on_run_end(self, **kwargs) -> None:
-        for _name, callback in self.callbacks.items():
-            callback.on_run_end(**kwargs)
+__all__ = ["Adversary"]
 
 
-class Attacker(AttackerCallbackHookMixin, torch.nn.Module):
-    """The attack optimization loop.
-
-    This class implements the following loop structure:
-
-    .. code-block:: python
-
-        on_run_start()
-
-        while true:
-            on_examine_start()
-            examine()
-            on_examine_end()
-
-            if not done:
-                on_advance_start()
-                advance()
-                on_advance_end()
-            else:
-                break
-
-        on_run_end()
-    """
+class Adversary(pl.LightningModule):
```

Comment:
Ideally this would inherit from `LitModular`. Then we can create pre-defined sequences for training.